### PR TITLE
Add ERC20Bridgeable: An ERC-7802 implementation

### PR DESCRIPTION
**Pull request by ernestognw**
### Description

This is an experiment to test how to use an ERC-7786 to create a gateway for cross-chain tokens following ERC-7802.

### CODE REVIEW

File: contracts/token/ERC20/extensions/draft-ERC20Bridgeable.sol [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    function _registerCrosschainERC20(string calldata caip2, string calldata crosschainERC20) internal virtual {
+        require(bytes(_crosschainERC20s[caip2]).length == 0, CrosschainERC20AlreadyRegistered(caip2));
+        _crosschainERC20s[caip2] = crosschainERC20;
+        emit RegisteredCrosschainERC20(caip2, crosschainERC20);
+    }
+
+    /// @dev Getter to check whether an attribute is supported or not.
+    function supportsAttribute(bytes4 selector) public view virtual returns (bool) {
+        return selector == crosschainMintAttr || selector == crosschainBurnAttr;
+    }
```
Comment (frangio) [inline contracts/token/ERC20/extensions/draft-ERC20Bridgeable.sol:72]: This is no longer needed. Also some events and variables at the top of the contract.

**Comment by ernestognw** [comment]
I'll test ERC-165 interface if we can agree on something around https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5350/files

**Comment by ernestognw** [comment]
Oh god

Captura de pantalla 2024-12-10 a la(s) 12 57 35 p m

File: contracts/mocks/token/draft-ERC20BridgeableMock.sol [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+abstract contract ERC20BridgeableMock is ERC20Bridgeable {
+    address bridge;
+
+    error OnlyTokenBridge();
+
+    event OnlyTokenBridgeFnCalled(address caller);
+
+    constructor(address bridge_, string memory name_, string memory symbol_) ERC20(name_, symbol_) ERC20Bridgeable() {
+        bridge = bridge_;
+    }
```
Comment (Amxx) [inline contracts/mocks/token/draft-ERC20BridgeableMock.sol:16]: we could just do 
```suggestion
    constructor(address bridge_) {
        bridge = bridge_;
    }
```
and the expose plugin will take care of the rest.

File: contracts/token/ERC20/extensions/draft-ERC20Bridgeable.sol [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        _;
+    }
+
+    /**
+     * @dev Checks if the caller is a trusted token bridge. MUST revert otherwise.
+     *
+     * Developers should implement this function using an access control mechanism that allows
+     * customizing the list of allowed senders. Consider using {Ownable}, {AccessControl} or {AccessManager}.
+     */
+    function _checkTokenBridge(address caller) internal virtual;
```
Comment (Amxx) [inline contracts/token/ERC20/extensions/draft-ERC20Bridgeable.sol:29]: usually we put these virtual not implemented function at the end of the contract.

File: test/token/ERC20/extensions/draft-ERC20Bridgeable.test.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        .withArgs(this.other.address, ethers.ZeroAddress, amount);
+
+      expect(await this.token.balanceOf(this.other.address)).to.equal(0);
+    });
+  });
+
+  describe('ERC165', function () {
+    shouldSupportInterfaces(['ERC7802'], {
+      ERC7802: ['crosschainMint(address,uint256)', 'crosschainBurn(address,uint256)'],
+    });
```
Comment (Amxx) [inline test/token/ERC20/extensions/draft-ERC20Bridgeable.test.js:84]: looking at this, I think we can do better on the `shouldSupportInterfaces`.

IMO here it should "just" be 
```
shouldSupportInterfaces({
    ERC7802: ['crosschainMint(address,uint256)', 'crosschainBurn(address,uint256)'],
});
```

Comment (Amxx) [reply test/token/ERC20/extensions/draft-ERC20Bridgeable.test.js:84]:
```suggestion
    shouldSupportInterfaces({
      ERC7802: ['crosschainMint(address,uint256)', 'crosschainBurn(address,uint256)'],
    });
```

File: contracts/crosschain/interfaces/draft-IERC7802.sol [hunk 5]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.27;
```
Comment (Amxx) [inline contracts/crosschain/interfaces/draft-IERC7802.sol:2]: any reason to put ^0.8.27 over ^0.8.20 ?

File: contracts/crosschain/interfaces/draft-IERC7802.sol
Comment (Amxx) [inline contracts/crosschain/interfaces/draft-IERC7802.sol:1]: should that really be in crosschain ?
It feels to me like it should either be in `contracts/token/extensions`, or just in `contracts/interfaces`

File: package.json [hunk 6]
```diff
@@ -49,28 +48,23 @@
     "@nomicfoundation/hardhat-chai-matchers": "^2.0.6",
     "@nomicfoundation/hardhat-ethers": "^3.0.5",
     "@nomicfoundation/hardhat-network-helpers": "^1.0.11",
+    "@openzeppelin/contracts": "file:lib/@openzeppelin-contracts",
```
Comment (Amxx) [inline package.json:51]: As a side effect, this fill get all the dependencies that are already present in the main repo.

That means we could technically remove anything we have in `devDependencies` that is not exclusive to the community repo. Do we want to do that ?

File: contracts/token/ERC20/extensions/draft-ERC20Bridgeable.sol [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    function crosschainBurn(address from, uint256 value) public virtual override onlyTokenBridge {
+        _burn(from, value);
+        emit CrosschainBurn(from, value, msg.sender);
+    }
+
+    /**
+     * @dev Checks if the caller is a trusted token bridge. MUST revert otherwise.
+     *
+     * Developers should implement this function using an access control mechanism that allows
+     * customizing the list of allowed senders. Consider using {Ownable}, {AccessControl} or {AccessManager}.
```
Comment (frangio) [inline contracts/token/ERC20/extensions/draft-ERC20Bridgeable.sol:48]: Ownable seems like an odd suggestion in this case.

```suggestion
     * customizing the list of allowed senders. Consider using {AccessControl} or {AccessManaged}.
```

**Review by Amxx — APPROVED** [review]

**Review by Amxx — APPROVED** [review]
If all tests pass, then that version is good for me.

Note: solhint will not be happy with some of the stuff in the mocks, but we'll see that when we enable it.

**Pull request merged**